lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs

```diff
@@ -164,25 +195,34 @@ mkTx
     -> (Address -> Maybe (k 'AddressK XPrv, Passphrase "encryption"))
     -> CoinSelection
     -> Either ErrMkTx (Tx, SealedTx)
-mkTx proxy (TxPayload certs mkExtraWits) timeToLive (rewardAcnt, pwdAcnt) keyFrom cs = do
+mkTx proxy pm (TxPayload certs mkExtraWits) timeToLive (rewardAcnt, pwdAcnt) keyFrom cs = do
     let withdrawals = mkWithdrawals
             proxy
             (toChimericAccountRaw . getRawKey . publicKey $ rewardAcnt)
             (withdrawal cs)
 
     let unsigned = mkUnsignedTx timeToLive cs withdrawals certs
 
-    addrWits <- fmap Set.fromList $ forM (CS.inputs cs) $ \(_, TxOut addr _) -> do
-        (k, pwd) <- lookupPrivateKey keyFrom addr
-        pure $ mkWitness unsigned (getRawKey k, pwd)
+    wits <- case (txWitnessTagFor @k) of
+        TxWitnessShelleyUTxO -> do
+            addrWits <- fmap Set.fromList $ forM (CS.inputs cs) $ \(_, TxOut addr _) -> do
+                (k, pwd) <- lookupPrivateKey keyFrom addr
+                pure $ mkShelleyWitness unsigned (getRawKey k, pwd)
+
+            let withdrawalsWits
```

Comment:
I was first wondering: why not factor out `fmap Set.fromList $ forM (CS.inputs cs) $ \(_, TxOut addr _)` from both cases, and just differentiate on the witnesses?

Then I see `withdrawalsWits` are only in the `TxWitnessShelleyUTxO` case. Wouldn't it _technically_ be possible (at least for people using this as a library) to spend both byron funds and withdrawals? Would it not be as well to construct `withdrawalsWits` in both cases then?

Reply:
that's good question - I have also this in the back of my head. I planned to have working byron/icarus tx...and then in the next PR check this possibility

Reply:
> to spend both byron funds and withdrawals?

Technically yes, but not with the wallet, because there are no withdrawals coming with our definition of Byron wallets.